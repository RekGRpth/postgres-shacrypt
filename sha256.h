/* Structure to save state of computation between the single steps.  */
struct sha256_ctx
{
  uint32_t H[8];

  uint32_t total[2];
  uint32_t buflen;
  char buffer[128];	/* NB: always correctly aligned for uint32_t.  */
};

/* Maximum salt string length.  */
#define SALT_LEN_MAX 16
/* Default number of rounds if not explicitly specified.  */
#define ROUNDS_DEFAULT 5000
/* Minimum number of rounds.  */
#define ROUNDS_MIN 1000
/* Maximum number of rounds.  */
#define ROUNDS_MAX 999999999

#define SHA256_SALT_PREFIX "$5$"
#define SHA256_ROUNDS_PREFIX "rounds="

char *sha256_crypt_r (const char *key, const char *salt, char *buffer, int buflen);

char *sha256_crypt (const char *key, const char *salt);

