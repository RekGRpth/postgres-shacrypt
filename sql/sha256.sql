select sha256_crypt('Hello world!', '$5$saltstring');
select sha256_crypt('Hello world!', '$5$rounds=10000$saltstringsaltstring');
select sha256_crypt('This is just a test', '$5$rounds=5000$toolongsaltstring');
select sha256_crypt('a very much longer text to encrypt.  This one even stretches over morethan one line.', '$5$rounds=1400$anotherlongsaltstring');
select sha256_crypt('we have a short salt string but not a short password', '$5$rounds=77777$short');
select sha256_crypt('a short string', '$5$rounds=123456$asaltof16chars..');
select sha256_crypt('the minimum number is still observed', '$5$rounds=10$roundstoolow');
