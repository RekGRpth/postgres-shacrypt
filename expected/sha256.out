select sha256_crypt('Hello world!', '$5$saltstring');
                       sha256_crypt                        
-----------------------------------------------------------
 $5$saltstring$5B8vYYiY.CVt1RlTTf8KbXBH3hsxY/GNooZaBBGWEc5
(1 row)

select sha256_crypt('Hello world!', '$5$rounds=10000$saltstringsaltstring');
                                 sha256_crypt                                 
------------------------------------------------------------------------------
 $5$rounds=10000$saltstringsaltst$3xv.VbSHBb41AL9AvLeujZkZRBAwqFMz2.opqey6IcA
(1 row)

select sha256_crypt('This is just a test', '$5$rounds=5000$toolongsaltstring');
                                sha256_crypt                                 
-----------------------------------------------------------------------------
 $5$rounds=5000$toolongsaltstrin$Un/5jzAHMgOGZ5.mWJpuVolil07guHPvOW8mGRcvxa5
(1 row)

select sha256_crypt('a very much longer text to encrypt.  This one even stretches over morethan one line.', '$5$rounds=1400$anotherlongsaltstring');
                                sha256_crypt                                 
-----------------------------------------------------------------------------
 $5$rounds=1400$anotherlongsalts$Rx.j8H.h8HjEDGomFU8bDkXm3XIUnzyxf12oP84Bnq1
(1 row)

select sha256_crypt('we have a short salt string but not a short password', '$5$rounds=77777$short');
                           sha256_crypt                            
-------------------------------------------------------------------
 $5$rounds=77777$short$JiO1O3ZpDAxGJeaDIuqCoEFysAe1mZNJRs3pw0KQRd/
(1 row)

select sha256_crypt('a short string', '$5$rounds=123456$asaltof16chars..');
                                 sha256_crypt                                  
-------------------------------------------------------------------------------
 $5$rounds=123456$asaltof16chars..$gP3VQ/6X7UUEW3HkBn2w1/Ptq2jxPyzV/cZKmF/wJvD
(1 row)

select sha256_crypt('the minimum number is still observed', '$5$rounds=10$roundstoolow');
                              sha256_crypt                               
-------------------------------------------------------------------------
 $5$rounds=1000$roundstoolow$yfvwcWrQ8l/K0DAWyuPMDNHpIVlTQebY9l/gL972bIC
(1 row)

